### PR TITLE
Feature/implement gitlab service

### PR DESCRIPTION
**Pull request by adelkahomolova**
## Description
Implement Gitlab Service and Gitlab Client.

## Motivation and Context

Gitlab Service - Support Gitlab paths.
Gitlab Client - Useful library doesn't exist.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] I have updated the documentation accordingly.
- [ ] I have added new practice to practice list in README.md.
- [x] I have read the **CONTRIBUTING** document.
- [x] I haven't repeated the code. (DRY)
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -11,39 +13,53 @@
...
-JavaScript/TypeScript | ✅
-Java/Kotlin | ✅
-Python | ✅
-PHP | 🚧
-C++ | 🚧
-C# | 🚧
-Ruby | 🚧
+dxs run https://gitlab.com/gitlab-org/gitlab
 
+dxs run https://bitbucket.org/atlassian/bitbucketjs
```
Comment (prokopsimek) [inline README.md]: Choose some smaller repo than these big ones. The user could copy&paste this row and try it and we should offer some faster scans.

File: src/services/git/__MOCKS__/gitLabServiceMockFolder/listIssuesResponse.ts [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+    },
+    url: 'https://gitlab.com/Homolova/ted_ontouml_kom/issues/1',
+    body: 'test',
+    createdAt: '2020-03-05T14:13:37.362Z',
+    updatedAt: '2020-03-05T14:13:37.362Z',
+    closedAt: null,
+    state: 'opened',
+    id: 1,
+  },
+];
```
Comment (prokopsimek) [inline src/services/git/__MOCKS__/gitLabServiceMockFolder/listIssuesResponse.ts]: Why is this here? Move to another file?

+ add type?

File: src/services/gitlab/GitLabService.ts [hunk 3]
```diff
@@ -0,0 +1,460 @@
...
+      totalCount = 10000;
+    }
+
+    return { totalCount, hasNextPage, hasPreviousPage, page, perPage };
+  }
+
+  /**
+   * Debug GitLab request promise
+   */
+  private unwrap<T>(clientPromise: Promise<CustomAxiosResponse<T>>): Promise<CustomAxiosResponse<T>> {
```
Comment (prokopsimek) [inline src/services/gitlab/GitLabService.ts]: You forget to implement this function to almost all Gitlab responses.

File: src/services/gitlab/GitLabService.ts [hunk 4]
```diff
@@ -0,0 +1,460 @@
...
+      });
+  }
+
+  /**
+   * Debug GitLab response
+   * - count API calls and inform about remaining rate limit
+   */
+  private debugGitLabResponse = <T>(response: CustomAxiosResponse<T>) => {
+    this.callCount++;
+    debug(`GitLab API Hit: ${this.callCount}. Remaining ${response.headers['RateLimit-Remaining']} hits.`);
```
Comment (prokopsimek) [inline src/services/gitlab/GitLabService.ts]: Has the Gitlab the header `RateLimit-Remaining` or any else?

File: src/services/gitlab/gitlabClient/GitLabClient.ts [hunk 5]
```diff
@@ -0,0 +1,31 @@
+import axios from 'axios';
+import { injectable } from 'inversify';
+
+@injectable()
+export class GitLabClient {
+  protected headers: { [header: string]: string };
+  protected host: string;
+  protected timeout: number;
+
+  constructor({ token, host = 'https://gitlab.com', timeout = 20000 }: ClientOptions) {
```
Comment (prokopsimek) [inline src/services/gitlab/gitlabClient/GitLabClient.ts]: 20 seconds as a timeout is quite high. I'd set it to 5 seconds to not block a user if something happens.

File: src/test/factories/responses/gitLab/issueResponseFactory.ts [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+      subscribed: false,
+      moved_to_id: null,
+      epic_iid: null,
+      epic: null,
+    },
+    params,
+  );
+};
+
+export const issueFromUser = {
```
Comment (prokopsimek) [inline src/test/factories/responses/gitLab/issueResponseFactory.ts]: Missing interface?

File: src/test/factories/responses/gitLab/listPrsResponseFactory.ts [hunk 7]
```diff
@@ -0,0 +1,6 @@
+import Bitbucket from 'bitbucket';
+import { MergeRequest } from '../../../../services/gitlab/gitlabClient/resources/MergeRequests';
+
+export const gitLabListPRsResponseFactory = (items: MergeRequest[] = []): MergeRequest[] => {
+  return items;
+};
```
Comment (prokopsimek) [inline src/test/factories/responses/gitLab/listPrsResponseFactory.ts]: Shouldn't be there a pagination?

File: src/test/factories/responses/gitLab/listPullCommitsResponseFactory.ts [hunk 8]
```diff
@@ -0,0 +1,5 @@
+import { Commit } from '../../../../services/gitlab/gitlabClient/resources/MergeRequests';
+
+export const gitLabListPullCommitsResponseFactory = (items: Commit[] = []) => {
+  return items;
+};
```
Comment (prokopsimek) [inline src/test/factories/responses/gitLab/listPullCommitsResponseFactory.ts]: Shouldn't be there a pagination?

File: src/test/factories/responses/gitLab/listRepoCommitsResponseFactory.ts [hunk 9]
```diff
@@ -0,0 +1,13 @@
+import { BitbucketCommit } from '../../../../services';
+import Bitbucket from 'bitbucket';
+
+export const bitbucketListCommitResponseFactory = (items: Bitbucket.Schema.Commit[]): BitbucketCommit => {
```
Comment (prokopsimek) [inline src/test/factories/responses/gitLab/listRepoCommitsResponseFactory.ts]: Bitbucket factory in Gitlab folder? (`src/test/factories/responses/gitLab/listRepoCommitsResponseFactory.ts`)

**Comment by prokopsimek** [comment]
@adelkahomolova It's impossible to evaluate Do PullRequests and Solve Issues Continuously practices.

```
$ ./bin/run run https://git.XXX.cz/dxheroes/YYY/identity-be -a TOKEN
WARNING: You're running DX Scanner in development mode.
  practices The Do PullRequests practice failed with this error:
  practices Error: Request failed with status code 404
  practices     at createError (/Users/prokop/Sites/_dxheroes/scanner/node_modules/axios/lib/core/createError.js:16:15)
  practices     at settle (/Users/prokop/Sites/_dxheroes/scanner/node_modules/axios/lib/core/settle.js:17:12)
  practices     at IncomingMessage.handleStreamEnd (/Users/prokop/Sites/_dxheroes/scanner/node_modules/axios/lib/adapters/http.js:236:11)
  practices     at IncomingMessage.emit (events.js:215:7)
  practices     at endReadableNT (_stream_readable.js:1183:12)
  practices     at processTicksAndRejections (internal/process/task_queues.js:80:21) +0ms
  practices The Solve Issues Continuously practice failed with this error:
  practices Error: Request failed with status code 404
  practices     at createError (/Users/prokop/Sites/_dxheroes/scanner/node_modules/axios/lib/core/createError.js:16:15)
  practices     at settle (/Users/prokop/Sites/_dxheroes/scanner/node_modules/axios/lib/core/settle.js:17:12)
  practices     at IncomingMessage.handleStreamEnd (/Users/prokop/Sites/_dxheroes/scanner/node_modules/axios/lib/adapters/http.js:236:11)
  practices     at IncomingMessage.emit (events.js:215:7)
  practices     at endReadableNT (_stream_readable.js:1183:12)
  practices     at processTicksAndRejections (internal/process/task_queues.js:80:21) +0ms
  practices The Do PullRequests practice failed with this error:
  practices Error: Request failed with status code 404
  practices     at createError (/Users/prokop/Sites/_dxheroes/scanner/node_modules/axios/lib/core/createError.js:16:15)
  practices     at settle (/Users/prokop/Sites/_dxheroes/scanner/node_modules/axios/lib/core/settle.js:17:12)
  practices     at IncomingMessage.handleStreamEnd (/Users/prokop/Sites/_dxheroes/scanner/node_modules/axios/lib/adapters/http.js:236:11)
  practices     at IncomingMessage.emit (events.js:215:7)
  practices     at endReadableNT (_stream_readable.js:1183:12)
  practices     at processTicksAndRejections (internal/process/task_queues.js:80:21) +0ms
  practices The Solve Issues Continuously practice failed with this error:
  practices Error: Request failed with status code 404
  practices     at createError (/Users/prokop/Sites/_dxheroes/scanner/node_modules/axios/lib/core/createError.js:16:15)
  practices     at settle (/Users/prokop/Sites/_dxheroes/scanner/node_modules/axios/lib/core/settle.js:17:12)
  practices     at IncomingMessage.handleStreamEnd (/Users/prokop/Sites/_dxheroes/scanner/node_modules/axios/lib/adapters/http.js:236:11)
  practices     at IncomingMessage.emit (events.js:215:7)
  practices     at endReadableNT (_stream_readable.js:1183:12)
  practices     at processTicksAndRejections (internal/process/task_queues.js:80:21) +0ms
```

File: src/services/gitlab/GitLabService.ts (hunk 4 above)
Comment (adelkahomolova) [reply src/services/gitlab/GitLabService.ts]: Yes, it has `RateLimit-Remaining` 
There are also values `RateLimit-Limit` and `RateLimit-Observed`

File: src/test/factories/responses/gitLab/listPrsResponseFactory.ts (hunk 7 above)
Comment (adelkahomolova) [reply src/test/factories/responses/gitLab/listPrsResponseFactory.ts]: As I've discussed test for pagination with you, there should be pagination but I'll add it in another PR, so I don't block this one.

File: package.json [hunk 10]
```diff
@@ -33,7 +33,7 @@
     "@types/qs": "6.9.1",
     "@types/table": "4.0.7",
     "@types/yaml": "1.2.0",
-    "axios": "0.19.2",
+    "axios": "^0.19.2",
```
Comment (prokopsimek) [inline package.json]: Do not unpin the version. It's important to have absolutely specific version.

File: README.md [hunk 11]
```diff
@@ -11,39 +13,53 @@
...
-JavaScript/TypeScript | ✅
-Java/Kotlin | ✅
-Python | ✅
-PHP | 🚧
-C++ | 🚧
-C# | 🚧
-Ruby | 🚧
+dxs run https://gitlab.com/ubon-refer/smart-refer-api
 
+dxs run https://bitbucket.org/maras333/node-api-with-ts
```
Comment (prokopsimek) [inline README.md]:
```suggestion
dx-scanner run https://bitbucket.org/maras333/node-api-with-ts
```

File: README.md [hunk 12]
```diff
@@ -11,39 +13,53 @@
...
   * [GitHub CI Action](#GitHub-Ci-Action)
 <!-- tocstop -->
+### Which version control system can you use?
 
-![DX Scanner Demo](./demo.svg)
+Sure you can use `GitHub`. You can also use `GitLab` and `Bitbucket` no matter if it is public or private repo!
 
+examples:
+```
+dxs run https://github.com/DXHeroes/dx-scanner
```
Comment (prokopsimek) [inline README.md]:
```suggestion
dx-scanner run https://github.com/DXHeroes/dx-scanner
```

File: README.md [hunk 13]
```diff
@@ -11,39 +13,53 @@
...
-Language | Supported
------------- | -------------
-JavaScript/TypeScript | ✅
-Java/Kotlin | ✅
-Python | ✅
-PHP | 🚧
-C++ | 🚧
-C# | 🚧
-Ruby | 🚧
+dxs run https://gitlab.com/ubon-refer/smart-refer-api
```
Comment (prokopsimek) [inline README.md]:
```suggestion
dx-scanner run https://gitlab.com/ubon-refer/smart-refer-api
```

File: src/scanner/Scanner.ts [hunk 14]
```diff
@@ -100,8 +100,13 @@ export class Scanner {
     this.d(`Components (${projectComponents.length}):`, inspect(projectComponents));
     const practicesWithContext = await this.detectPractices(projectComponents);
     this.d(`Practices (${practicesWithContext.length}):`, inspect(practicesWithContext));
-    await this.fix(practicesWithContext);
-    const practicesAfterFix = await this.detectPractices(projectComponents);
+
+    let practicesAfterFix;
```
Comment (prokopsimek) [inline src/scanner/Scanner.ts]: Is it `any` either after line `107`?

File: src/services/gitlab/GitLabService.ts [hunk 15]
```diff
@@ -0,0 +1,468 @@
...
+        repo: {
+          url: `${this.host}/${owner}/${repo}`,
+          name: repo,
+          id: data.project_id.toString(),
+          owner: user,
+        },
+      },
+    };
+
+    //TODO
```
Comment (prokopsimek) [inline src/services/gitlab/GitLabService.ts]: TODO what? Please describe what's necessary to do.

File: src/scanner/Scanner.ts (hunk 14 above)
Comment (adelkahomolova) [reply src/scanner/Scanner.ts]: It can be PracticeWithContext[] or undefined, will fix it

File: src/services/gitlab/GitLabService.ts (hunk 15 above)
Comment (adelkahomolova) [reply src/services/gitlab/GitLabService.ts]: Sorry, I just forgot about it there.
